src/go/metrics/gauge.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 The Magma Authors.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"time"
+
+	"github.com/benbjohnson/clock"
+)
+
+type Gauge interface {
+	// Set sets the Gauge to an arbitrary value.
+	Set(value float64, labels map[string]string)
+
+	// GetSamples returns all values set since it was last called.
+	GetSamples() <-chan *GaugeSample
+
+	Name() string
+
+	Description() string
+
+	SetClock(clock clock.Clock)
+}
+
+type GaugeSample struct {
+	Value       float64
+	LabelValues map[string]string
+	TimestampMs int64
+}
+
+type gaugeImpl struct {
+	name        string
+	description string
+	labelNames  []string
+	samples     chan *GaugeSample
+	clock       clock.Clock
+}
+
+// NewGauge creates a new Gauge.
+// It is expected that values will be populated not more than once a minute.
+func NewGauge(name string, description string, labelNames []string) Gauge {
+	gauge := &gaugeImpl{name, description, labelNames, make(chan *GaugeSample, 10), clock.New()}
+	RegisterGauge(gauge)
+	return gauge
+}
+
+// Set sets the Gauge to an arbitrary value.
+func (g *gaugeImpl) Set(value float64, labels map[string]string) {
```

Comment:
if `g.samples` is a `make(chan *GaugeSample, 10)`, it means after the 10th call to `Set` without draining the channel, the caller will block (almost definitely unintended)

we are probably better off with a straight mutex array here:

```
type gaugeImpl struct {
  ... other metadata ...
  samples []GaugeSample  // note, a plain struct is fine here; explicit pointer is not necessary since these are not nil-able and copies should be small/light
  sync.Mutex
}

func (g *gaugeImpl ) Set(value float64, labels map[string]string) {
  g.Lock()
  defer g.Unlock()

  // note: labels is a pointer to a map, so if the caller is sharing the labels map and changing it, the samples will all get the same copy. probably ok to note that in the documentation somewhere; making defensive copies would be safer, but incur a large perf penalty
  g.samples = append(s.samples, GaugeSample{...})
}

func (g *gaugeImpl) GetMetric() *io_prometheus_client.Metric {
  samples := g.samples
  g.Lock()
  g.samples = nil
  g.Unlock()

  // create *io_prometheus_client.Metric and return
}
```

---

if we really want to do channels that pipeline into a result, then we'd want to feed the data directly into a `*metricsd.MetricsContainer` via a goroutine inside `Collector`. Then every `period`, a mutex pauses the collector goroutine, swaps out the `*metricsd.MetricsContainer` for a new one, and sends the previous filled-up `*metricsd.MetricsContainer` to `Sender`

this is a much more advanced design and probably overkill; it also risks creating pauses in callers to `Gauage.Set` since it'll be governed by the collector goroutine.

Reply:
Oops. Thought I forgot something. Intended for the following:
- `Set` to just delete old values if they aren't drained - though perhaps this should be a configurable parameter
- channel size to be adjustable, with defaults based on how frequently metrics are collected (every minute), and the expectation that each metric is set at most once per minute, per label combination

I'll rework this bit.